packages/typespec-test/test/contentsafety_modular/generated/typespec-ts/review/ai-content-safety.api.md

```diff
@@ -129,6 +133,9 @@ export interface PagedTextBlocklist {
 export interface RemoveBlockItemsRequestOptions extends OperationOptions {
 }
 
+// @public (undocumented)
+export function setContinuationToken(page: unknown, continuationToken: string | undefined): void;
```

Comment:
@xirzec I notice [your comment](https://github.com/Azure/autorest.typescript/issues/1199#issuecomment-1289735138) in original issue, do you think it's better to re-design below two helpers or just expose the same in modular?

Currently we just expose the get(set)ContinuationToken.

https://github.com/Azure/autorest.typescript/blob/9642e55f45349bb18b20fc2b026b62c1069b6663/packages/typespec-ts/test/modularIntegration/pageable.spec.ts#L60-L79

Reply:
I support us figuring out a better abstraction here and perhaps getting rid of the 'maxPageSize' param as well -- maybe we need to rethink the core-paging contract for modular?

@deyaaeldeen - what do you think? Is this borrowing too much trouble to fix now? I feel like if we don't tackle it for the new DPG codegen we're going to kick this can forever down the road.

Reply:
@xizec regarding getting rid of this `maxPageSize` parameter in core, my understanding is if the relevant tsp defines the `maxPageSize` as query parameter, we should also respect this parameter in codegen side and so that customers could also set their own values, right?

```typescript
const iter = client.list({ 
       maxpagesize: 3 
});
```

Reply:
> @deyaaeldeen - what do you think? Is this borrowing too much trouble to fix now? I feel like if we don't tackle it for the new DPG codegen we're going to kick this can forever down the road.

That is a good point! I gave this some thought and I wonder if [this proposal](https://www.typescriptlang.org/play?target=99#code/JYOwLgpgTgZghgYwgAgApwOYQMoTGUDAZ2QG8AoZZBAe3FAFc4C6AVGgawhAH4AuZETBRCAbkrIAtnAAe6LNmAAvCP2QgGkgEbRxAX3LlQkWIhTzQzCABMAgkQCeIBAEkTcLQBsIAHgmsAUW9JbjByAD5kCBlIEGsSeydXdy9fQODQyLIJLQd5CAEfVnzkAF5kdIgQ8ABtAF0AGgr83HxCEnKWvAIQYnCACiJu9rVizBxh3qIASjLIxOc3aA9vH1JkAAdxgTGsJtp6DWZgNk5uASERXuQ9cP0gA) makes sense. The paging type variables get relegated to `byPage` itself emphasizing that it is not a core scenario to use `byPage` and assuming most scenarios don't require to know the paging details. It also updates the `byPage`'s return type to return both the page and the continuation token, making the relationship between both explicit and close by.

Reply:
I think we can safely remove maxPageSize if it's not working currently and it should be considered as non-breaking. 

For continuationToken, from the HLC  
![image](https://github.com/Azure/autorest.typescript/assets/55688292/4af8f10d-f969-4d1a-8bae-ec47ce2c772d)
 It seems to me continuationToken = nextLink, and the setContinuationToken 
https://github.com/qiaozha/azure-sdk-for-js/blob/70507137e6626c4ed33d72bb59ba109ce0b9dc02/sdk/advisor/arm-advisor/src/pagingHelper.ts#L13 
just stores the continuationToken for each page data, I wonder if that's necessary? Is there a real scenario that customers want to getContinuationToken from where they have list those data before. If they want, they could just pass the nextLink to byPage as the continuationToken?

Reply:
@qiaozha the requirement of `getContinuationToken` is coming from customers' ask and you could refer [their code here](https://github.com/Azure/autorest.typescript/issues/1199#issuecomment-1310817741). The intention is because we hide the nextLink detail in our public API so we provide two utilities to help them extract the underlying tokens in byPage iterator.

@deyaaeldeen I like the idea of the return type change for `byPage` and compared with current helper way and we could have two benefits:
- page items and its continuationToken are highly coupled so that we could encapsulate them in one place
- this could be easier to discover and use for customers

The drawback for relegating paging type variables is that we can't pass customized `PageSettings` when using getPagedAsyncIterator. These types could only be passed during calling byPage, which means they can't be used in codegen/SDK side but only for customers.

I think the `maxPageSize` issue still exists and directly throwing exception is not acceptable for me. Could you help me understand if getting ride of this in core would be chanllenging or not? Because I notice this param could be valuable for other downstreams. e.g [service bus](https://github.com/azure/azure-sdk-for-js/blob/61be11030937904bdc689fd4d0384e53ea250b33/sdk/servicebus/service-bus/src/serviceBusRuleManager.ts#L218-L219).

Reply:
> I think the maxPageSize issue still exists and directly throwing exception is not acceptable for me

I think we can take a code gen approach to supporting `maxPageSize` instead of always including it. If the API supports a `top` parameter (designated as such by a typespec decorator?), then generate a `maxPageSize` option and use it to set `top`. If the API doesn't, then don't generate such a parameter.

> The drawback for relegating paging type variables is that SDK would lose the ability to pass customized PageSettings when using getPagedAsyncIterator. These types could only be passed during calling byPage and only for customers.

If we're using the code gen approach to `PageSettings` described above, this shouldn't be an issue because `PageSettings` will no longer be a type variable. We can also use a code gen approach to the `TPage` type and get ride of the type variable as well. So for an API that doesn't have a `top` and has a page type of `Collection`, code gen can generate the following:
```ts
interface PageSettings {
  continuationToken?: string;
}

interface PaginatedAsyncIterable<
  TElement
> extends AsyncIterable<TElement>  {
  byPage: (settings?: PageSettings) => AsyncIterable<{ page: Collection, continuationToken: string }>;
}
```

Reply:
Yeah I think we could take a code gen approach. Regarding the `maxPageSize` I would prefer to get rid of this in PageSettings with following reasons:

- We already expose the option `maxPageSize` in operation-level option which means we could 1) either duplicate it or 2) remove the operation one. The former would introduce confusion and the latter would cause real breakings for HLC.
- On the other hand just getting rid of this causes the API surface breaking but it would be safe considering the inner is directly throwing exception. And all operations could share the same `PageSettings` 
- There is no clear mapping from the real option to maxPageSize in tsp, it depends on the service team's implementation. I know some services use the `top`, some are `maxresults` and others use `maxpagesize`